types.go

```diff
@@ -17,72 +17,151 @@ limitations under the License.
 package fission
 
 type (
-	// Metadata is used as the general identifier for all kinds of
-	// resources managed by the controller.
-	Metadata struct {
-		Name string `json:"name"`
-		Uid  string `json:"uid,omitempty"`
+	//
+	// Functions and packages
+	//
+
+	// Checksum of package contents when the contents are stored
+	// outside the Package struct. Type is the checksum algorithm;
+	// "sha256" is the only currently supported one. Sum is hex
+	// encoded.
+	Checksum struct {
+		Type string `json:"type"`
+		Sum  string `json:"sum"`
 	}
 
-	// Function is a unit of executable code.  Though it's called
-	// a function, the code may have more than one function; it's
-	// usually some sort of module or package.
-	Function struct {
-		Metadata    `json:"metadata"`
-		Environment Metadata `json:"environment"`
-		Code        string   `json:"code"`
+	// PackageType is either literal or URL, indicating whether
+	// the package is specified in the Package struct or
+	// externally.
+	PackageType string
+
+	// Package contains or references a collection of source or
+	// binary files.
+	Package struct {
+		// Type defines how the package is specified: literal or URL.
+		Type PackageType `json:"type"`
+
+		// Literal contents of the package. Can be used for
+		// encoding packages below TODO (256KB?) size.
+		Literal []byte `json:"literal"`
+
+		// URL references a package.
+		URL string `json:"url"`
```

Comment:
I have not read through all the changes yet, but why do you need the distinction here between literal and URL? That might be something to extract from the actual `Package` struct to something like this:
```
PackageLocation (Url/Filesystem) ---fetch--> Package
```
That would make the package agnostic of what way it was retrieved.